### DISCUSSION TITLE
YouTube paste function for .md

### DISCUSSION THREAD

**Discussion by allreddv** [linkarzu/dotfiles-latest#5] [Ideas]
Hey, linkarzu.  

Thanks for all your videos, you have really helped dial in my nvim setup, especially for note taking in .md  

I made this really cool function I thought you might like but thought it was probably too big to place as a YT comment. This function allows you to paste a YT url from clipboard and then it will use the cli tool yt-dlp to do an api call and extract the meta data for the video in json,  I have it setup to paste the url and in the text section for the url link in .md and it places the Channel Name, Video Title, and upload date.  It then adds the thumbnail image on the next line.  I like to have access to the thumbnail images because it helps me remember the video easier as i am scrolling through videos I save for topics in notes.  

  Two things to note: 
  1. you will have to install yt-dlp that is an open source cli tool for downloading YT videos and/or it's audio, but you can also skip that and just get the meta data which is what I am using it for.  Make sure to install yt-dlp and not yt-dl as the later has been deprecated.  
  2. It takes about 10 seconds to complete the action.  So if it looks like it isn't working, just wait.  I have tried different methods of getting the json to see if I could make it quicker, but nothing changed.  Even trying to just call for the bits of info I wanted didn't speed it up, so that is why I just stuck with getting the full meta data json.  You can pull whatever out of it that you want to adjust to your liking.  For me, I just wanted channel, title, and upload date(to know time relevancy)  but you can get anything you want like view counts, comment counts, etc.  You can also adjust the function to do the same thing for an entire playlist if you wanted because yt-dlp has the functionality built in for it. 
  
  Here is a screenshot of what it looks like when I copy an YT link to clipboard and then paste in .md file using just my keybinding <leader>mpy   I have my keys set like <leader>".md related" "pasting urls" "YT" because i have other url pasting functions also depending on the site I want to paste from. 

![Screenshot 2024-10-25 at 11 20 49 AM](https://github.com/user-attachments/assets/204d37a5-d4d7-40c3-b486-8575bd96812c)

Here is the function,  Let me know if you like it. I have been using it a lot and have not had any errors or issues on any of the urls I have pasted. 

```
-- Helper function to convert camelCase to spaced words
local function camel_to_spaces(str)
	return str:gsub("(%l)(%u)", "%1 %2")
end

-- Function to extract YouTube metadata and insert it into the markdown file
function _G.insert_youtube_metadata()
	-- Get the URL from the clipboard
	local url = vim.fn.getreg("+")

	-- Check if the URL is valid
	if not url:match("youtube.com/watch") then
		vim.notify("No valid YouTube URL in clipboard", vim.log.levels.ERROR)
		return
	end

	-- Run yt-dlp asynchronously to extract metadata
	local cmd = string.format("yt-dlp_macos -J --skip-download '%s'", url)
	vim.fn.jobstart(cmd, {
		-- Buffer output for complete JSON parsing
		stdout_buffered = true,
		on_stdout = function(_, data, _)
			-- Concatenate and filter out empty lines from the data
			local output = table.concat(data, "")
			if output == "" then
				return
			end

			-- Use Neovim's JSON parsing
			local ok, metadata = pcall(vim.json.decode, output)
			if not ok or not metadata then
				vim.notify("Failed to parse YouTube metadata", vim.log.levels.ERROR)
				return
			end

			-- Extract channel name, video title, upload date, and thumbnail
			local channel_name = metadata.uploader or "Unknown Channel"
			local video_title = metadata.title or "Unknown Title"
			local upload_date = metadata.upload_date or "Unknown Date"
			local thumbnail = metadata.thumbnail or ""

			-- Format the upload date to YYYY-MM-DD
			if upload_date:match("%d%d%d%d%d%d%d%d") then
				upload_date = upload_date:sub(1, 4) .. "-" .. upload_date:sub(5, 6) .. "-" .. upload_date:sub(7, 8)
			end

			-- Convert camel case to spaces for channel and title
			channel_name = camel_to_spaces(channel_name)
			video_title = camel_to_spaces(video_title)

			-- Format the markdown output
			local markdown_output = string.format(
				"[ YouTube - %s - %s - %s](%s)\n![Thumbnail](%s)\n",
				channel_name,
				video_title,
				upload_date,
				url,
				thumbnail
			)

			-- Insert the formatted metadata at the current cursor position
			vim.api.nvim_put(vim.split(markdown_output, "\n"), "l", true, true)
		end,

		on_stderr = function(_, data, _)
			if #data > 0 then
				vim.notify("Error extracting metadata: " .. table.concat(data, "\n"), vim.log.levels.ERROR)
			end
		end,

		on_exit = function(_, exit_code, _)
			if exit_code ~= 0 then
				vim.notify("yt-dlp exited with error", vim.log.levels.ERROR)
			end
		end,
	})
end

-- Set up the keybinding
vim.keymap.set("n", "<leader>mpy", _G.insert_youtube_metadata, { desc = "Insert YouTube Metadata" })

```

**Comment by allreddv** [comment]
I started to occasionally get an error message for failure to extract metadata even though the function was still working and pasting correctly.  So I started to look into it and it had to do with the on_stderr callback function.  So I fixed it, now not only do i not get errors again, it processes everything much faster and now it will retrieve the metadata and paste all the info in about 1-2 seconds.

the only thing that needs to be changed in the above function is the on_stderr callback at the bottom.  Just needs to be changed to this. 

		on_stderr = function(_, data, _)
			-- Only display stderr output if it contains critical errors
			local error_message = table.concat(data, "\n")
			if error_message:find("ERROR") then
				vim.notify("Error extracting metadata: " .. error_message, vim.log.levels.ERROR)
			end
		end,

**Reply by linkarzu** [reply]
Really appreciate it, I'll keep it in mind when I need a keymap similar to what yours does, seems pretty useful!
